### PR TITLE
feat(generator): add parent awareness to union values

### PR DESCRIPTION
**Pull request by jsenko**
## Summary

Move parent/metadata to the `Any` interface so ALL values in the tree (including non-entity union values like `BooleanUnionValue`) track their parent node. This fixes 57 `NodePath` resolution failures where union values had `_parent == null`.

Part of #1041. Unblocks the JSON Schema spec rewrite.

## Changes

- `Any.java` — add `parent()`, `parentPropertyName()`, `parentPropertyType()` with default null
- `UnionValueImpl.java` — add `_parent`, `_parentPropertyName`, `_parentPropertyType` fields + setters
- `CreateImplMethodsStage` — generated setters and collection add methods now set parent/metadata on all union values (not just entity variants)

## How it works

Before: `node.setNot(booleanUnionValue)` → setter checks `isEntity()` → false → no parent set → `booleanUnionValue.parent()` returns null

After: `node.setNot(booleanUnionValue)` → setter checks `isEntity()` → false → falls to else branch → sets `_setParent(this)`, `_setParentPropertyName("not")` → `booleanUnionValue.parent()` returns the owning entity

## Test plan

- [x] All 1129 project tests pass
- [x] Synthetic snapshot compiles
- [x] FullGeneratorTest compiles all generated files

### CODE REVIEW

**Review by EricWittmann — COMMENTED** [review]
## Code Review

**+655 / -213** across 34 files — thorough and well-structured change.

### Overview

The design is sound — lifting parent/metadata tracking from `Node`-only to the `Any` interface so that non-entity union values (e.g. `BooleanUnionValue`) can also track their parent. The generator changes correctly cover setters, add/insert, remove/clear, and factory methods for all property shapes (standard, list, map, union).

### Positive observations

- **Good API design** — default methods on `Any` mean existing implementations don't break (only `NodeImpl` and `UnionValueImpl` need to override).
- **Comprehensive coverage** — the generator handles all union subtypes: entity, entity-list, entity-map, and the new else-branch for primitive/non-entity union values.
- **Clean removal of `attach()`** — the old `attach()` required the parent to already be attached, which was fragile. The new approach of automatically setting parent in generated setters/add methods is much better.
- **Reader order fix** is correct — calling `addXxx()` before `readXxx()` ensures the parent is set before recursive child reading. Subtle but important.
- **`DeleteAllChildSchemasCommand` reorder** is correct — `addOldSchema(s)` now runs before `removeAnyOfSchema()` since remove now calls `detach()`.
- The Javadoc on `Any` is excellent — nice documentation of the attachment lifecycle.
- `NodeImpl.root()` null-safety fix is a good defensive change independent of the union work.

### Issues

**1. `remove` methods detach unconditionally (potential bug)**

In `createRemoveMethodBody` for lists, `detach()` is called before confirming the value is actually in the collection:
```java
if (value != null) value.detach();
this.${fieldName}.remove(value);
```
If `value` is not actually in this collection (e.g. caller passes a node that belongs to a different parent), it will still be detached from that other parent — corrupting the other tree. Should be guarded:
```java
if (value != null && this.${fieldName}.remove(value)) {
    value.detach();
}
```
The map `removeItem` doesn't have this issue since `map.remove()` returns the removed value which is checked for null.

**2. No `detach()` on the old value when a setter overwrites**

When `setFoo(newValue)` is called and `this.foo` already holds an old value, the old value is not detached — it retains a stale parent reference. This is pre-existing behavior, but it's now more consequential since union values also track parents. Worth a follow-up if not addressed here.

**3. Double `_setParent()` in union entity-list reader**

In `CreateReadersStage.java`, the reader explicitly calls `((NodeImpl) model)._setParent(node)` on each entity in a temporary list. When the union value is later set via `node.setItems(unionValue)`, the generated setter iterates the inner list and calls `_setParent()` again on each entity. Harmless (same parent both times), but worth a comment in the generator explaining why it's needed here (model goes into a temp list, not directly through an add method).

**4. `attach()` removal is a breaking public API change**

`Node.attach(Node parent)` was public interface API. Removing it will break any external consumers. The test files show two internal callers migrated to `((NodeImpl) rval)._setParent(...)`, which requires a downcast to the concrete class. Worth calling out in release notes or considering whether a public re-parent method should exist on `Any` or `Node`.

**5. Test callers use raw cast to `NodeImpl`**

In `DereferenceTestReferenceResolver.java` and `ValidationTest.java`:
```java
((io.apitomy.datamodels.models.NodeImpl) rval)._setParent(from.parent());
```
These hard-cast to the concrete `NodeImpl` class using a fully-qualified name. Previously `attach()` was on the `Node` interface. Consider whether `_setParent()` should be exposed on `Node` for internal/test usage to avoid the cast.

### Verdict

Architecture is solid and the generator coverage is comprehensive. Issue #1 (unconditional detach on remove) is the most actionable — it can silently corrupt a different tree if a caller passes a non-member value. The others are follow-up material. The synthetic test snapshots and the passing 1129 tests give good confidence this works correctly for the normal paths.

**Comment by EricWittmann** [comment]
## CI Failure Analysis

Both CI jobs failed for the same root cause. Java 17 failed in the **Build Project** step; Java 21 was cancelled as a consequence.

**All Java tests pass** (1129 tests, 0 failures). The failure is in the **TypeScript/npm test suite**, which runs after JSweet transpilation.

### Root Cause

Two TypeScript test files reference the removed `attach()` method on `Node`:

```
FAIL tests/dereference.test.ts
  ● Test suite failed to run
    tests/dereference.test.ts:55:14 - error TS2339: Property 'attach' does not exist on type 'Node'.
    55         rval.attach(from.parent());

FAIL tests/validation.test.ts
  ● Test suite failed to run
    tests/validation.test.ts:75:14 - error TS2339: Property 'attach' does not exist on type 'Node'.
    75         rval.attach(from.parent());
```

The PR updated the **Java** test files (`DereferenceTestReferenceResolver.java` and `ValidationTest.java`) to use `((NodeImpl) rval)._setParent(from.parent())` instead of `rval.attach(from.parent())`, but the corresponding **TypeScript** test files (`tests/dereference.test.ts` and `tests/validation.test.ts`) were not updated. Since `attach()` was removed from the `Node` interface, the TS compiler rejects the call.

### Fix

Update the two TypeScript test files to replace `rval.attach(from.parent())` with the equivalent `_setParent` call (or whatever the TS-side equivalent is after JSweet transpilation).

**Test result summary:** 2 failed, 6 passed, 8 total (npm test suites)

**Comment by jsenko** [comment]
Good catch! Fixed by updating the two TS test files to use `(rval as any)._setParent(from.parent())` instead of `rval.attach(from.parent())`. CI is now green.
